### ISSUE TITLE
createPipelineAsync() doesn't reject its promise, even if the created pipeline is invalid

### ISSUE THREAD

**Issue by litherum** [gpuweb/gpuweb#3296]
From https://gpuweb.github.io/gpuweb/#dom-gpudevice-createcomputepipelineasync:

> When pipeline is ready to be used or has been made [invalid](https://gpuweb.github.io/gpuweb/#invalid), [resolve](https://webidl.spec.whatwg.org/#resolve) promise with pipeline.

It seems like we should be rejecting the promise if the created pipeline is invalid. I don't see any reason not to, and it would help alleviate some confusion about "how do you know if the pipeline created with `createPipelineAsync()` ended up being valid or not?"

**Comment by kainino0x** [comment]
I'm pretty sure we discussed this, before but I would have to dig to figure out where. I probably proposed this because it is more symmetrical with the non-async createPipeline methods. The sync versions won't interrupt the flow of your JS when they fail. I'm not sure whether it's a good idea to have the async versions do so. This is related to #1629 somewhat.

On the other hand, with the current spec language, if you do want to detect errors, then the createPipelineAsync functions have to report to error scopes. You can't wrap an error scope around the resolution of the promise, so we'd need it to report errors into the error scope that was surrounding the createPipelineAsync call (which may be popped now). This is actually fine with the editors' proposal to #2119 although I'm not clear anymore whether it was accepted or not. I guess we didn't finish discussing it.

---

I still kind of want to remove these functions in service of #2085. We briefly discussed doing so in #2583, but rejected that approach. If we did so, the ambiguity would be resolved.

**Comment by kainino0x** [comment]
Well, the issue wouldn't be entirely resolved. We could still reject the promise. But it would resolve the ambiguity of whether the createPipelineAsync methods report to the error scope or not.

Anyway, I've filed #3297.

**Comment by Kangz** [comment]
A couple things to take into account for this issue:

 - Do we foresee compilation messages being emitted on pipelines in the future? A promise rejection would prevent getting them (since we'd have no object returned to JS to anchor the messages on).
 - Given that lost devices should behave as much as possible as non-lost devices, should lost devices never reject an async pipeline compilation but return an error pipeline? Or on the contrary should they always reject (and applications are expected to handle that correctly?)

**Comment by litherum** [comment]
Model 1: (please forgive my bad javascript)

```
pushErrorScope();
let promise1 = createPipelineAsync();
let promise2 = popErrorScope();
await let pipeline, errorScope = Promise.all([promise1, promise2]);
if (errorScope == undefined) {
    // use pipeline
} else {
    // handle the error
}
```

Model 2:

```
createPipelineAsync().then(function(pipeline) {
    // use pipeline
}, function(errorScope) {
    // handle the error
});
```

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/gpu-web-2022-09-07-08-apac-timed#createpipelineasync-doesnt-reject-its-promise-even-if-the-created-pipeline-is-invalid-3296">GPU Web 2022-09-07/08 APAC-timed</a></summary>

* MM: if you use createPipelineAsync you get a Promise, and it's resolved even if it's a compile failure. Might expect rejection in that case because that's what you'd reject for. Understand why we might not do this synchronously. Infrastructure's all set up to use. Why not use it?
* BJ: reminds me of WebXR issue - had Promise resolving, you'd ask, does my system support this mode? Promise coming back, originally resolved/rejected depending on support. Got feedback from web platform folks (W3C) that this is an antipattern. Had request - always resolves to a boolean. Logic: we asked a question. If it could provide an answer - it's functioning correctly. Rejecting should be used in an exceptional case - the thing you requested has failed ot happen. Not sure which bucket I'd put this in. You're saying, yes, we tried to do the thing and it failed - in all other parts of the API we try to keep trucking along. I'd probably do that. Can look at it as - you asked for a thing and we couldn't provide an answer for it. These are reasons why you might want to reject in stead of resolve. Since we return valid-looking objects in other cases, I'm inclined to do the same here.
* MM: example you gave from WebXR, makes sense to be a boolean resolve rather than reject. In this case it's pretty clear it's not comparable - app says do something, and we couldn't do that. Analogy: fetch. Please download this file; if it can't, the Promise rejects. Here, we can't compile the shader.
* KN: don't think that's how Fetch works. 404 - Fetch will resolve, not reject.
* KG: which question are you asking? Give me a successfully compiled thing, or a compilation attempt to be made?
* KN: couple arguments from the issue in favor of keeping it as-is, and not rejecting.
* KN: makes it more symmetric with other createPipeline methods.
* KN: we already have the notion of async errors in WebGPU. Other cases where we reject when this happens; but it's more symmetric with this API.
* createRenderPipeline -> createRenderPipelineAsync - more symmetric if we don't reject, but return an invalid pipeline.
* Corentin's argument - if we put compilation / error messages on the pipelines, we'll need an object to put those on. Technically, could reject with that object - would have to duplicate that API to do that if we had a rejection. If we don't reject we can put it on the pipeline object.
* KN: I think these are strong enough arguments to keep things as is.
* MM: second argument - it's natural to reject with a reason why it failed. Maybe I don't agree with how invasive this factoring would be.
* KN: we'll still need to give you that info when you create it synchronously. A function you call on the pipeline to give you the info object. You could get a pipeline object from the async creation path.
* BJ: shader modules right now - no way to fetch errors, but can fetch messages, including deprecation warnings. Want these even in success case. Seems weird to bifurcate this given you want this for the synchronous case as well.
* KG: one idea - feels terrible but checks some boxes - when we accept, we give you a pipeline. What if we also give you a pipeline with the rejection? Won't work, but you can ask it for messages. Would let you await compileshadermodule / createpipelineasync - get it back and expect it to work. Alternatively, like try/await. If you don't get the right thing we'll limp along. Have a clear "here's the good pipeline".
* MM: sounds reasonable to me.
* BJ: not aware of anywhere else on the platform that has a pattern like that. Typically when you reject there's an amorphous object coming out of it, but it's usually an exception object.
* KN: it's like throwing the pipeline object, which we try not to do.
* DG: in web standards folks, wanting a boolean, shouldn't we give back the pipeline object and let you query it?
* MM: would this task have completed successfully if you tried it? One's a boolean, one's a task.
* JB: two different mechanisms for permitting async/streaming behavior. Stream a bunch of requests- not equivalent to async code. Just writing things with awaits everywhere, you have to await for completion before submitting next command. There isn't a parity between these. Both are ways to permit computation to keep on going. Seems to me that in ways where we report errors, we signal exceptions eagerly. Whole point of providing async entry points is that the caller says, I don't mind waiting for you. There's more info available at that time. The motivation for other entry points to return invalid objects doesn't apply to the async operations, and I think we should give people all the information they have, in the way they're expecting.
* KN: think its' best to think about what's practical for developers rather than what's philosophical meaning of asking for a pipeline asynchronously. Right now, don't have a way to ask whether a pipeline's valid, even asynchronously. To know, you have to put ErrorScope around pipeline creation and check its result. This is something we can still do with async creation. You'd use the ErrorScope around the createPipelineAsync call, and it'd report the error there. Things do get tricky. If you put ErrorScope around an await, it can catch other stuff due to the await. Might be an argument that we should reject; that similarity with synchronous API might be undesirable for this reason.
* BJ: KG you mentioned the web platform folks might ask us to things a certain way anyway. True. I can see both sides' arguments. **Maybe reach out to web platform mentors in our various orgs and ask? Have a design issue, what's the webby thing to do here?** Have had good success asking this.
* KG: sounds good - let's all of us ask this.
* DG: reject / resolve - as the API user, if we reject, allow us to move error handling - refactor code. Get resolved promise - can use without error checking, and if rejected - handle that there and return an error code. Something to consider.
* KN: we should consider the programming pattern in these two cases.
* JB: I don't understand the rationale for these async functions in the first place.
* MM: 1) i wrote what I think the programming model should be today - not great.
* MM: reason for async functions - if you didn't have them, you wouldn't know when your pipeline was ready. will take multiple frames. don't want to createPipeline then draw, will require it to be compiled this frame. want to let user wait until it's ready.
* KG: feels like we can write down more things to look at this from more sides.
* KN: i'll send a link in a moment - talks about why we have this.
    * [https://github.com/gpuweb/gpuweb/issues/3297](https://github.com/gpuweb/gpuweb/issues/3297)
* DG: I'd appreciate that.

</details>

**Comment by kainino0x** [comment]
I agree that the example for model 1 is bad. Model 2 is definitely better. I'm going to propose another idea, but if people don't like it, I think I'm fine with model 2.

---

I still think we should avoid mixing two different async programming models (WebGPUs and JavaScript's) in this way, and we could do better than model 2. I would like to possibly consider this proposal yet again (sorry for continually coming back to it): https://github.com/gpuweb/gpuweb/issues/3297#issuecomment-1211322491
BUT with a tweak: disallow using a pipeline (in a command buffer) until it's "ready". This would avoid actually changing the semantics of the API, while also massaging the async models together better.

- Remove the `create*PipelineAsync` functions.
- In `GPUPipelineDescriptorBase` add `boolean reportReady = false;`.
- In `GPUPipelineBase` add `readonly attribute Promise<undefined> ready;` which resolves when ready if requested, or otherwise is an already-resolved promise.
- In `setPipeline` add validation that the pipeline has already reported that it's ready (the promise has already resolved). Most likely, throw an exception in this case, because that simplifies implementation (the service side doesn't need to create a pipeline object until it's ready).

The trickiest bit of this is probably that errors would still need to be routed to the error scope that surrounded the `create*Pipeline` call (like in model 1) even though they're intended to be generated later than that.

This design would prep better for the possibilities of: forcing wait by using a pipeline that isn't ready (#3297) and surfacing "compilation messages" on pipelines (https://github.com/gpuweb/gpuweb/issues/3296#issuecomment-1238339693).

**Comment by kainino0x** [comment]
The example above for this proposal:

```js
device.pushErrorScope();
const pipeline = createPipeline({ reportReady: true });
// It would be an error to use the pipeline here, because it hasn't reported ready.
const error = await device.popErrorScope();
// You've awaited, so now undefined whether you can use the pipeline
// (same as if you awaited anything else). We could make this more well defined.
await pipeline.ready;
// Now you can use the pipeline.
```

**Comment by kainino0x** [comment]
The value is that validation can prevent usage of the pipeline before it's ready. This way implementations aren't required to implement the "join" semantics for using a pipeline before it's ready (#3297), which is not easy.

Ken raised the concern that without the "join" semantics, usage of the pipeline will be flaky due to JS async timing, and applications could easily accidentally rely on timing of a particular implementation. Given this, I'm in favor of reject-on-failure.

**Comment by greggman** [comment]
So, now that error reporting is unordered (#3362) do we even need `createPipelineAsync`?

Why can't I just call `createPipeline` (which is always async and only stalls on usage), and if I want to know if it's valid and successful before I try to use it I wrap it in `pushErrorScope`/`popErrorScope`?

Then there's only one error reporting mechanism and a simpler API.

(Sorry if already forgot why this was NG 😅)

**Comment by kainino0x** [comment]
There are (were) 2 reasons:
- Implementations may use createPipelineAsync as a signal to create the pipeline on a background thread. Dawn for example will createPipeline on the "main" thread, as we don't currently implement the "join" semantics needed to run it in the background. We could signal this in another way, but it would again require the "join" semantics.
    > I think part of the difficulty here is that we've been discussing this issue as if there are two modes (sync and async), when in fact there are three: foreground thread, background thread with reportback, and background thread without reportback.
    > -- https://github.com/gpuweb/gpuweb/issues/3297#issue-1332648756


- popErrorScope doesn't officially guarantee that the pipeline is ready - only that we were done checking for "validation" errors. However, now that we also have ["internal"](https://gpuweb.github.io/gpuweb/#dom-gpuerrorfilter-internal) error scopes, those probably would guarantee it.

**Comment by greggman** [comment]
Thank you for explanation.

I guess I question whether this "signal" to compile on a background thread is a reasonable signal. I'm guessing the implementation is free to ignore this signal? It can compile sychronously and just return an already resolved promise.

It would seem like a reasonable optimization for an implementation to compile all pipelines on background threads. In that sense, the signal is superfluous. The implementation can do the optimal thing always. There's even an incentive for it to do so as it should lead to faster startup times for apps. It can also do the optimal thing never (ignore the signal).

Given that, it still seems like `createPipelineAsync` is not needed. An optimal implementation will background compile all pipelines. If the user wants to prevent stalls, they wrap the creation in an error scope.

Is the "join" that difficult?

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/gpu-web-2022-10-05#createpipelineasync-doesnt-reject-its-promise-even-if-the-created-pipeline-is-invalid-3296">GPU Web 2022-10-05</a></summary>

* KN: KG, the plan was for you to put together a proposal.
* KG: one more week please.
* CW: we can wait another week but the reject option seems good enough. I think we should aim for a high bar of good enough. There will be quirks in WebGPU anyway. Rejection could have some quirks but it's fine. What we have already is fine too. Simple change. Bigger changes are worse.
* KR: last week Myles said he was convinced by my flakiness argument.
* KN: KG, aside from interest in seeing your proposal, he didn't say he wouldn't be happy doing this.
* CW: let's try to gain consensus offline if we can.

</details>

**Comment by litherum** [comment]
@kdashg IIRC last time we discussed this in the call, you said you were going to come up with a proposal. Is there anything I can help with?

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/gpu-web-meeting-2022-10-12-13-apac-timed#createpipelineasync-doesnt-reject-its-promise-even-if-the-created-pipeline-is-invalid-3296">GPU Web Meeting 2022-10-12/13 APAC-timed</a></summary>

* JB: general agreement that sync/async distinction isn't great indicator about pipeline creation on-thread vs. off-thread
* JB: browser can always take one or the other impl. Browser can always move off-thread if it wants.
* JB: lot of discussion in issue - what tactics do WebGPU users actually want to use when they're trying to meet a framerate, but can't guarantee pipeline creation in advance? Trying to introduce pipelines in the middle of play.
* JB: if they have fallback strategy, can use the fallback pipeline, and switch to the new one if created.
* If they don't have a fallback - their only strategy is to jank. Miss frames until pipeline's ready.
* Talking with different folks about what devs actually want to use - gotten different people saying different things. Overall - game engines would prefer to have fallbacks, but often don't. Jank when they can't get things ready.
* Our concern - seems to us that games most often jank. They don't want to render until they can draw everything. They'd like to miss frames. This is the actual practice in the industry. This is a use case WebGPU should make sure it supports.
* In the proposal right now - once we decide to use async, we have no way to jank, if we want to. Either await it, or use it and get an error. Can't say, just pause the whole queue until it's ready.
* We're concerned that there's this use case common in the industry that we're not facilitating.
* KR: This is why the two entry points exist. Developers can use the synchronous version if they want to jank when the pipeline is used. Async is very compatible with JavaScript. We are working with developers loading massive data who want to be able to avoid jank. The two entry points facilitate two use cases.
* JB: once someone chooses to use the async one, they have given up the ability to jank.
* KR: They can use the non-async version instead
* KG: one concern - when you have a case you're trying to load a pipeline, want to do work after it's loaded, but may need to say - I need to render - want to jank - you can do that in the current design. Call createPipeline again with the same descriptor. Not ideal. Feel a bit bad about it. It's a use case that feels strange. Wish you could createPipeline with the option to have a promise. But in the grand scheme of things we can accept - what I expect to happen - in the future we do add a Promise to GPUPipeline, so in the future you can await on a pipeline.
* In that case - maybe slightly vestigial createPipelineAsync. Like a polyfill around createPipeline and awaiting ready(). Probably OK.
* Feels like an acceptable outcome to me. Not super happy about it.
* Think we should tolerate the compromise and accept the cruft.
* JB: seconding what Kelsey said. We understand Google's under a lot of pressure. This is where we expect feedback based on impl experience. We expect our feedback will be handled by the committee.
* KR: Understood. My personal standpoint isn’t coming from pressure to ship, but from the fact we’ve had this entry point for years now and it was well reasoned. Structure of API designed so you can do both. I’m not sure what’s going to happen if we add a promise to the pipeline and what developers are going to do. Whether they’ll inadvertently jank vs understanding properly how to use the promise. Based on what developers have done - Babylon may use this for example. Want to base this decision on the experience we already have.
* KG: agree that it has value and is well-reasoned. Just think this could be even better.
* KN: fully agree that we may end up reaching a point where createPipelineAsync is cruft. I don't know whether we are or not. Need to prototype, work with developers.
    * KG: I'm proposing we don't do that work.
    * KN: not arguing with that.
* KN: I agree that we may end up in that future, and I strongly support keeping what we have right now rather than weakly supporting, but based on the same reasoning.
* KG: wrote [https://hackmd.io/aZvyrbEzSsO3r6homVh4kQ](https://hackmd.io/aZvyrbEzSsO3r6homVh4kQ)
* KG: figuring out what I would have wanted. Comparison, number of use cases. Not pursuing that direction any more. Interesting, wanted to share that exploration.
* JB: Moz saying overall - we're fine with proceeding with this question. Take the change where we reject the Promise. Otherwise the API stays the same. Just wanted to make sure the alternatives we were pursuing are properly recorded.
* KG: do we have consensus?
* MM: which consensus?
* KG: createPipelineAsync will reject Promise if pipeline creation fails.
* MM: OK. Reject with nothing? Reject with errors? Editor discussion?
* JB: **we can make that a separate PR**.
* MM: I prefer to let the editors handle this.
* KG: **would like it if the editors would bring a proposal here**.
* MM: OK. **Right now, rejects with nothing. Then CG will build on this**.

</details>

**Comment by litherum** [comment]
Resolved this week: Reject the promise (with nothing) if compilation failed.

**Comment by litherum** [comment]
I have pretty serious concerns about the "with nothing" clause above, because if the promise is rejected with nothing, the author won't get any error messages, which means they won't know what went wrong.

I think we have to at least resolve it with error messages...

**Comment by kainino0x** [comment]
We'll definitely need to iron the rejection value. @litherum I already have #3523 open which just rejects with an OperationError. We have to reject with some kind of Error type according to some W3C guidance docs, so there will always be an error message string. There's some other problems though, see what I posted in #3523.

**Comment by litherum** [comment]
I opened https://github.com/whatwg/webidl/issues/1223 to ask what the idiomatic way of doing this is.

**Comment by kainino0x** [comment]
This has been agreed upon and will be closed by #3523, I just need to wrap that up and land it.
